### PR TITLE
feat(eks): fargate profiles

### LINKED ISSUE #5303: feat(aws-eks): Fargate support
https://aws.amazon.com/blogs/aws/amazon-eks-on-aws-fargate-now-generally-available/

### Other

https://github.com/aws-cloudformation/aws-cloudformation-coverage-roadmap/issues/288

* [ ] :wave: I may be able to implement this feature request
* [ ] :warning: This feature might incur a breaking change

---

This is a :rocket: Feature Request

### PR DESCRIPTION
**Pull request by eladb**
Add support for Fargate profiles. Profiles can be added to a cluster through the `addFargateProfile` method or by defining a `FargateProfile` construct.

The EKS custom resource was extended to support a new resource type that provisions Fargate profiles and slightly refactored to allow maximum code reuse between the two resource types.

The IAM policy for the cluster role was extended to allow creation of Fargate Profiles (`eks:CreateFargateProfile`, `eks:DescribeFargateProfile`, `eks:DeleteFargateProfile`, `iam:GetRole`, and `iam:CreateServiceLinkedRole`). Since we don't know the name of the profile/cluster during synthesis, the IAM policy is wider than desired (uses `*` for resource ARNs).

In order to enable Fargate-only cluster, the CoreDNS inside the Kubernetes cluster must be patched to use "fargate" as the compute type for CoreDNS. To that end, an additional resource, `KubernetesPatch`  was added to the kubctl custom resource provider. This resource type takes "apply" and "restore" manifests which are patched upon CREATE and DELETE, respectively.

The `FargateCluster` high level construct is a sugar for defining a Fargate-only EKS cluster.

Resolves #5303 

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by eladb** [comment]
Copy @pahud

File: packages/@aws-cdk/aws-eks/lib/fargate-profile.ts [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+      throw new Error(`Fargate profile requires at least one selector`);
+    }
+
+    if (props.selectors.length > 5) {
+      throw new Error(`Fargate profile supports up to five selectors`);
+    }
+
+    this.tags = new TagManager(TagType.MAP, 'AWS::EKS::FargateProfile');
+    for (const [ key, value ] of Object.entries(props.tags || {})) {
+      this.tags.setTag(key, value);
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-eks/lib/fargate-profile.ts]: Why not leave this up to the aspect-based tagger?

File: packages/@aws-cdk/aws-eks/lib/aws-auth-mapping.ts [hunk 2]
```diff
@@ -1,4 +1,6 @@
-
+/**
+ * AwsAuth mapping.
+ */
 export interface Mapping {
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-eks/lib/aws-auth-mapping.ts]: Don't know that `Mapping` is a good name for it, a little too generic for the purpose it seems to serve.

File: packages/@aws-cdk/aws-eks/lib/kubectl-handler/apply/__init__.py [hunk 3]
```diff
@@ -56,7 +56,7 @@ def kubectl(verb, file):
             output = subprocess.check_output(cmd, stderr=subprocess.STDOUT)
         except subprocess.CalledProcessError as exc:
             output = exc.output
-            if 'i/o timeout' in output and retry > 0:
+            if b'i/o timeout' in output and retry > 0:
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-eks/lib/kubectl-handler/apply/__init__.py]: OMG!

File: packages/@aws-cdk/aws-eks/lib/cluster-resource.ts [hunk 4]
```diff
@@ -67,13 +63,34 @@ export class ClusterResource extends Construct {
...
+        : '*'
+    });
+
     this.creationRole.addToPolicy(new iam.PolicyStatement({
-      actions: [ 'eks:CreateCluster', 'eks:DescribeCluster', 'eks:DeleteCluster', 'eks:UpdateClusterVersion', 'eks:UpdateClusterConfig' ],
+      actions: [ 'eks:CreateCluster', 'eks:DescribeCluster', 'eks:DeleteCluster', 'eks:UpdateClusterVersion', 'eks:UpdateClusterConfig', 'eks:CreateFargateProfile' ],
       resources: [ resourceArn ]
     }));
 
+    this.creationRole.addToPolicy(new iam.PolicyStatement({
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-eks/lib/cluster-resource.ts]: Does it make sense to do this only if and when `addFargateProfile` is called?

File: packages/@aws-cdk/aws-eks/lib/cluster.ts [hunk 5]
```diff
@@ -601,6 +651,47 @@ export class Cluster extends Resource implements ICluster {
...
     tagAllSubnets('public', this.vpc.publicSubnets, "kubernetes.io/role/elb");
   }
+
+  /**
+   * Patches the CoreDNS deployment configuration and sets the "eks.amazonaws.com/compute-type"
+   * annotation to either "ec2" or "fargate". Note that if "ec2" is selected, the resource is
+   * omitted/removed, since the cluster is created with the "ec2" compute type by default.
+   */
+  private defineCoreDnsComputeType(type: CoreDnsComputeType) {
+    if (!this.kubectlEnabled) {
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-eks/lib/cluster.ts]: Is this worth an error?

File: packages/@aws-cdk/aws-eks/lib/fargate-profile.ts (hunk 1 above)
Comment (eladb) [reply packages/@aws-cdk/aws-eks/lib/fargate-profile.ts]: Are you proposing not to expose `tags` as a property?

File: packages/@aws-cdk/aws-eks/lib/aws-auth-mapping.ts (hunk 2 above)
Comment (eladb) [reply packages/@aws-cdk/aws-eks/lib/aws-auth-mapping.ts]: That's a good point. Let me deprecate it.

Comment (eladb) [reply packages/@aws-cdk/aws-eks/lib/aws-auth-mapping.ts]: Done

File: packages/@aws-cdk/aws-eks/lib/kubectl-handler/apply/__init__.py (hunk 3 above)
Comment (eladb) [reply packages/@aws-cdk/aws-eks/lib/kubectl-handler/apply/__init__.py]: Yeah...

File: packages/@aws-cdk/aws-eks/lib/cluster.ts (hunk 5 above)
Comment (eladb) [reply packages/@aws-cdk/aws-eks/lib/cluster.ts]: it's a private API, but ok, I've wrapped the outer call with a condition and added an error here.

File: packages/@aws-cdk/aws-eks/lib/fargate-profile.ts (hunk 1 above)
Comment (rix0rrr) [reply packages/@aws-cdk/aws-eks/lib/fargate-profile.ts]: Yes?

Comment (eladb) [reply packages/@aws-cdk/aws-eks/lib/fargate-profile.ts]: OK. I'll remove. Seems like that's the common pattern (we should awslint this shit).

**Pull request merged**